dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/BackwardsReachingDefinitionsAnalysis.java

```diff
@@ -0,0 +1,388 @@
+        return result;
+    }
+
+    /**
+     * Lists all potential users of the result of an event.
+     * @param writer Event of interest to fetch information about.
+     * @throws IllegalArgumentException If {@code writer} was not inside the program when it was analyzed.
+     */
+    public Readers getReaders(RegWriter writer) {
+        Preconditions.checkNotNull(writer, "writer is null");
+        final Readers result = writerMap.get(writer);
+        Preconditions.checkArgument(result != null, "writer %s has not been analyzed.", writer);
+        return result;
+    }
+
+    /**
+     * Lists all potential users of uninitialized registers.
+     */
+    public Readers getInitialReaders() {
+        final Readers result = writerMap.get(null);
+        Preconditions.checkState(result != null, "initial state has not been analyzed.");
+        return result;
+    }
+
+    /**
+     * Analyzes an entire function definition.
+     * @param function Defined function to be analyzed.  Its parameters are marked as initialized.
+     */
+    public static BackwardsReachingDefinitionsAnalysis forFunction(Function function) {
+        final var analysis = new BackwardsReachingDefinitionsAnalysis();
+        final Set<Register> finalRegisters = new HashSet<>();
+        analysis.initialize(function, finalRegisters);
+        analysis.run(function, finalRegisters);
+        analysis.postProcess();
+        return analysis;
+    }
+
+    /**
+     * Analyzes an entire set of threads.
+     * @param program Contains a set of threads to be analyzed.  Additionally-defined functions are ignored.
+     * @param branch Optional, queried for possibility of pairs of writers appearing together in an execution.
+     */
+    public static BackwardsReachingDefinitionsAnalysis forProgram(Program program, BranchEquivalence branch) {
+        final var analysis = new BackwardsReachingDefinitionsAnalysis();
+        final Set<Register> finalRegisters = finalRegisters(program);
+        for (Function function : program.isUnrolled() ? program.getThreads() :
+                Iterables.concat(program.getThreads(), program.getFunctions())) {
+            analysis.initialize(function, finalRegisters);
+            analysis.run(function, finalRegisters);
+        }
+        analysis.postProcess();
+        if (branch != null && program.isUnrolled()) {
+            analysis.analyzeMust(branch);
+        }
+        return analysis;
+    }
+
+    public static BackwardsReachingDefinitionsAnalysis injectForProgram(Program program, Context analysisContext) {
+        final BranchEquivalence branch = analysisContext.get(BranchEquivalence.class);
+        final BackwardsReachingDefinitionsAnalysis analysis = forProgram(program, branch);
+        analysisContext.register(BackwardsReachingDefinitionsAnalysis.class, analysis);
+        analysisContext.register(ReachingDefinitionsAnalysis.class, analysis);
```

Comment:
Why do you need to register the concrete class here? If not needed, I would prefer to move the registering to calling code so as to not break the pattern we have.

Reply:
It is indeed not required anywhere and can be removed.